src/pipeline/src/tablename.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use dyn_fmt::AsStrFormatExt;
+use regex::Regex;
+use snafu::{ensure, OptionExt};
+use yaml_rust::Yaml;
+
+use crate::error::{Error, InvalidTableNameTemplateSnafu, RequiredTableNameTemplateSnafu, Result};
+use crate::{PipelineMap, Value};
+
+const REPLACE_KEY: &str = "{}";
+
+lazy_static::lazy_static! {
+    static ref NAME_TPL: Regex = Regex::new(r"\$\{([^}]+)\}").unwrap();
+}
+
+/// TableNameTemplate is used to generate target table name from pipeline context.
+/// The config should be placed at the end of the pipeline.
+/// Use `${variable}` to refer to the variable in the pipeline context, the viarable can be from input data or be a processed result.
+/// Note the variable should be an integer number or a string.
+/// In case of any error occurs during runtime, the input default table name will be used.
+///
+/// ```yaml
+/// tablename: a_${xxx}_${b}
+/// ```
+///
+/// For example, if the template is `a_${xxx}_${b}`, and the pipeline context is
+/// `{"xxx": "123", "b": "456"}`, the generated table name will be `a_123_456`.
+#[derive(Debug, PartialEq)]
+pub(crate) struct TableNameTemplate {
+    pub template: String,
+    pub keys: Vec<String>,
+}
+
+impl TableNameTemplate {
+    pub fn apply(&self, val: &PipelineMap) -> Option<String> {
+        let values = self
+            .keys
+            .iter()
+            .filter_map(|key| {
+                let v = val.get(key)?;
+                match v {
```

Comment:
Should we add boolean conversion?

Reply:
I think it's strange for a table name to be something like `normal_name_true`. We can add it later if there is a request.